### ISSUE TITLE
v0.9.7 pha-zuellig sale issue

### ISSUE THREAD

**Issue by joaoluis-pdm** [PharmaLedger-IMI/fgt-workspace#93]
See TST logs v0.9.7 on
https://homes.pdmfc.com/jpsl/PharmaLedger/phazuellig20220725.zip

Problems with a sale around pha-zuellig.log
2022-07-25T13:22:11.132906436Z

User attempts to perform a sale and gets 
```txt
2022-07-25T13:22:13.329173689Z [FGT-API - 1658755333329] - _sendResponse= {
2022-07-25T13:22:13.329186278Z   status: 400,
2022-07-25T13:22:13.329188728Z   error: 'Bad Request',
2022-07-25T13:22:13.329190858Z   message: 'Could not insert record with id string on table sales'
2022-07-25T13:22:13.329194968Z }
```
(but it is followed by a message to an MAH with a sale afterwards).

Then attempts to perform the sale again but fails with 
```txt
2022-07-25T13:22:41.167756889Z [FGT-API - 1658755361167] - _sendResponse= {
2022-07-25T13:22:41.167764319Z   status: 400,
2022-07-25T13:22:41.167766459Z   error: 'Bad Request',
2022-07-25T13:22:41.167768449Z   message: 'Product gtin: 74310288032194, batchNumber: V6264H, serialNumber: 201 already sold.'
2022-07-25T13:22:41.167770569Z }
```

@pccosta-pdm  can you figure out what is going in by the log, or do you want to schedule a meeting to ask for more details ?

**Comment by pccosta-pdm** [comment]
The issue was because we created a DSU for each list of products that we grouped by MAH, this process was used to notify the MAHs of the sale and later inserted into the DB.

When trying to insert a record in the DB with an ID that already exists, the error occurred, but the notifications had already been sent to the suppliers.

Sale request flow:
> (1) sale request 
> (2) check stock availability (and check with MAH if the product has already been sold)
> (3) **create sale dsu (by mah) and notify mah** 
> (4) insert DB record 
> (5) commit batch

in a nutshell:
**1st try:** create a sale with an id already taken with productX (that equals to gtin+batch+serialNumber) > check stock availability (sends a notification to MAH to check if the product has already been sold) (2) > create dsu and notify mah (3) > mah add the product as already sold > insert db record (4) > **error because sale id already exists in db**

**2nd try**: create a sale with a new id with productX (that equals to gtin+batch+serialNumber) > check stock availability (sends a notification to MAH to check if the product has already been sold)  > **error because mah notifies pha that productX has already sold**

the fix updates the sale request flow to:
> sale request 
> check stock availability 
> **create sale dsu (by mah)** 
> insert DB record 
> commit batch 
> **notify mah's**

**Comment by joaoluis-pdm** [comment]
@TiagoV-PDMFC  do you have any chance of reviewing the previous fix proposal ? (@pccosta-pdm  after lunch please explain me, that the 'Could not insert record with id string on table sales' error is because a duplicate sale is being attempted, right ?)

**Comment by joaoluis-pdm** [comment]
@pccosta-pdm but do please fix the missing schema in swagger /sale/create

![image](https://user-images.githubusercontent.com/77448470/181785035-ed35c1a5-d037-46c2-b87c-26b7e30e7790.png)

**Comment by pccosta-pdm** [comment]
> @pccosta-pdm but do please fix the missing schema in swagger /sale/create
> 
> ![image](https://user-images.githubusercontent.com/77448470/181785035-ed35c1a5-d037-46c2-b87c-26b7e30e7790.png)

I created issue #105 to solve this issue and others related to making a sale (UI only)

**Comment by joaoluis-pdm** [comment]
Closed, as #105 handles the UI side.
